fix(webrtc example): clarify idle connection timeout

DougAnderson444 opened the pull request:
## Description

When I ran the `example/browser-webrtc` example I discovered it would break after a ping or two.
The `Ping` idle timeout needed to be extended, on both the server and the wasm client, which is what this PR fixes.
I also added a small note to the README about ensuring `wasm-pack` is install for the users who are new to the ecosystem.

Fixes: #4950.

## Notes & open questions

The native code can accept `u64::MAX`, but curiously in `wasm` we cannot use `u64::MAX`. The `with_idle_connection_timeout` breaks in the browser on anything bigger than `32_212_254u64`

```rust
.with_swarm_config(|c| {
  c.with_idle_connection_timeout(Duration::from_secs(
   32_212_254u64,
   // 32_212_255u64, // breaks, too large for wasm?
  ))
})
```

The reason why this occurs is left as an open challenge for the community.

## Change checklist

<!-- Please add a Changelog entry in the appropriate crates and bump the crate versions if needed. See <https://github.com/libp2p/rust-libp2p/blob/master/docs/release.md#development-between-releases>-->

- [x] I have performed a self-review of my own code
- [x] I have made corresponding changes to the documentation
- ~~[ ] I have added tests that prove my fix is effective or that my feature works~~
- ~~[ ] A changelog entry has been made in the appropriate crates~~

DougAnderson444 committed:
set timeout to slightly greater than Ping interval

The example was timing out because the idle connection time was 5 seconds and the ping interval is 15s

DougAnderson444 committed:
add install note for `wasm-pack` to README

it's the small things that help out new users

DougAnderson444 committed:
set idle timeout to u64::MAX

Since Ping does not `KeepAlive`, we set the idle timeout to a very large number of seconds

DougAnderson444 committed:
extend the idle timeout on the server

DougAnderson444 committed:
set wasm32 target idle value to maximum

For some reason, the largest value that works in wasm32 target (in Chrome) is 32_212_254u64.

Everything larger than this breaks in the browser. Cause unknown at this point.

DougAnderson444 committed:
Merge branch 'libp2p:master' into fix-webrtc-example

thomaseizinger commented on examples/browser-webrtc/src/main.rs:
```diff
@@ -38,7 +38,7 @@ async fn main() -> anyhow::Result<()> {
         .with_behaviour(|_| ping::Behaviour::default())?
         .with_swarm_config(|cfg| {
             cfg.with_idle_connection_timeout(
-                Duration::from_secs(30), // Allows us to observe the pings.
```
Why is 30 not enough to observe some pings?

DougAnderson444 replied:
I was trying to be consistent with the `Ping` example here:

https://github.com/libp2p/rust-libp2p/blob/c1925e5e612bd609e57d6ac50a7b28cf16a1aa2c/examples/ping/src/main.rs#L42

But see my comment below, I've tweaked the example to be more clear I think

thomaseizinger commented on examples/browser-webrtc/src/lib.rs:
```diff
@@ -24,7 +24,9 @@ pub async fn run(libp2p_endpoint: String) -> Result<(), JsError> {
         })?
         .with_behaviour(|_| ping::Behaviour::new(ping::Config::new()))?
-        .with_swarm_config(|c| c.with_idle_connection_timeout(Duration::from_secs(5)))
+        // Ping does not KeepAlive, so we set the idle connection timeout to 32_212_254u64,
```
Why such a large value? Isn't 30 seconds entirely sufficient to see that the connection works?

DougAnderson444 replied:
Part of this was trying to determine / document why there was a strange idle limit in wasm in case someone tries to do `u64::MAX` like I initially tried to do. For the example though I made some changes which clear it up

thomaseizinger reviewed (commented):
Two comments. Also linking a related discussion / proposal: https://github.com/libp2p/rust-libp2p/pull/4967.

DougAnderson444 committed:
add verbose 30s beginning, duration, and end

Added more clarity about how long the Ping is intended to last, and made it a bit more obvious that the ConnectionClosed event is expected in the case of `Ping` after the idle timeout expires

thomaseizinger commented on examples/browser-webrtc/src/lib.rs:
```diff
@@ -46,6 +51,19 @@ pub async fn run(libp2p_endpoint: String) -> Result<(), JsError> {
+                    body.append_p("All done with pinging! ")?;
+
+                    break;
+                } else {
```
This `else` is not needed because you `break` before (which I think is good style!)

DougAnderson444 replied:
Ah, good eyes! I'll fix this.

DougAnderson444 committed:
rm the `else` after break;

as per @thomaseizinger 's comment

DougAnderson444 committed:
Merge branch 'master' into fix-webrtc-example

thomaseizinger reviewed (approved):
Great changes! Thank you very much :)

If you are keen, I'd be interested in your opinion on the wider topic of idle timeouts in https://github.com/libp2p/rust-libp2p/pull/4967.

mergify[bot] committed:
Merge branch 'master' into fix-webrtc-example
